services/terrain/src/terrain/persistence/search.clj

```diff
@@ -110,46 +110,55 @@
       (throw+ es-uninitialized))))
 
 
-(defn- tags-access-filter
-  [tags memberships]
-  (letfn [(tag-filter [tag] (query/term :id {:type  "tag"
-                                             :id    tag
-                                             :path  "targets.id"
-                                             :cache false}))]
-    (query/bool :must   (query/nested :path   "userPermissions"
-                                      :filter (query/term "userPermissions.user" memberships))
-                :should (map tag-filter tags))))
+(defn- mk-filter
+  [in-folders tags memberships]
+  (let [filter-path (fn [path] (query/prefix "path" (if (= \/ (last path))
+                                                      path
+                                                      (str path \/))))
+        filter-tag  (fn [tag] (query/term :id {:type  "tag"
+                                               :id    tag
+                                               :path  "targets.id"
+                                               :cache false}))
+        perm-filter (query/nested :path   "userPermissions"
+                                  :filter (query/term "userPermissions.user" memberships))]
+    (query/bool :must   (query/bool :must perm-filter :should (map filter-path in-folders))
+                :should (map filter-tag tags))))
```

Comment:
The elasticsearch docs are a little confusing. I had to test it out manually to make sure that `must` and `should` work together as I hoped.  The `must` and `must not` clauses must always pass.  If there is a should `clause` at least `minimum_should_match` or all of entries should pass for a result to be returned.

The `minimum_should_match` parameter defaults to `1`.